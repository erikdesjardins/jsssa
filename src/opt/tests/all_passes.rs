macro_rules! case {
    ( $name:ident, $string:expr ) => {
        #[test]
        fn $name() -> Result<(), crate::err::NiceError> {
            use crate::{ast2ir, ir, opt, parse, swc_globals};
            swc_globals::with(|g| {
                let (ast, _) = parse::parse(g, $string)?;
                let ir = ast2ir::convert(g, ast);
                let ir = opt::run_passes(g, ir);
                let ppr = ir::print(g, &ir);
                insta::assert_snapshot_matches!(stringify!($name), ppr, $string);
                Ok(())
            })
        }
    };
}

case!(
    precompute,
    r#"
    x = 1 + 1 + 1 + 1;
    y = "foo" + " " + "bar";
"#
);

case!(
    downleveling,
    r#"
    let x = 1;
    x = 2;
    x = 3;

    let y = 10;
    log(y);
    log(y + 1);
"#
);

case!(
    downleveling_bail,
    r#"
    let x = 1;
    x = 2;
    x = 3;
    if (foo) log(x);

    let y = 10;
    log(y);
    log(y + 1);
    if (bar) (function() {
        y = 5;
    })();
"#
);
