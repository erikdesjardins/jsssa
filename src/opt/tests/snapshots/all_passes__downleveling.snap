---
created: "2019-04-02T05:19:14.444045900Z"
creator: insta@0.7.1
source: src/opt/tests/all_passes.rs
expression: "\n    let x = 1;\n    x = 2;\n    x = 3;\n\n    let y = 10;\n    log(y);\n    log(y + 1);\n"
---
_ini = 10
_fun = <global log>
<dead> = _fun(_ini)
_fun$1 = <global log>
_arg = 11
<dead> = _fun$1(_arg)
